File: solutions/ex3_rotation.f90
program rotation
use mpi_f08
implicit none
integer, parameter :: n=64
integer i,proc,nproc,prev,next,iter
type(MPI_Status) :: status
integer, allocatable :: A(:)
call MPI_Init()
call MPI_COMM_RANK(MPI_COMM_WORLD, proc)
call MPI_COMM_SIZE(MPI_COMM_WORLD, nproc)
if (proc==0 .and. (mod(n,nproc) /= 0 .or. mod(n,2) /= 0)) then
   print *, 'Number of processes incompatible with the size'
   call MPI_Abort(MPI_COMM_WORLD,1)
endif
allocate (A(0:n/nproc))
do i=1,n/nproc
   a(i)=proc
enddo
print '(A,i2,A,64i5)', 'proc=',proc,' iter=0 A=',A(1:n/nproc)
do iter=1,n
  if (proc == 0) then
     prev=nproc-1
  else
     prev=proc-1
  endif
  if (proc==nproc-1) then
     next=0
  else
     next=proc+1
  endif
  if (mod(proc,2)==0) then
     call MPI_Send(A(n/nproc),1,MPI_INTEGER,next,iter,MPI_COMM_WORLD)
     call MPI_Recv(A(0),1,MPI_INTEGER,prev,iter,MPI_COMM_WORLD,status)
  else
     call MPI_Recv(A(0),1,MPI_INTEGER,prev,iter,MPI_COMM_WORLD,status)
     call MPI_Send(A(n/nproc),1,MPI_INTEGER,next,iter,MPI_COMM_WORLD)
  endif
  do i=n/nproc,1,-1
     A(i)=A(i-1)
  enddo
  print '(A,i2,A,i3,A,64i5)', 'proc=',proc,' iter=',iter,' A=',A(1:n/nproc)
enddo
deallocate (A)
call MPI_Finalize()
end program rotation

File: solutions/ex3_rotation.c
#include <stdio.h>
#include "mpi.h"
#include <stdlib.h>
#define n 64

int main(int argc, char *argv[])
{

int i,ierr,proc,nproc,prev,next,iter;
MPI_Status status;
int *A;

ierr = MPI_Init(&argc,&argv);

ierr = MPI_Comm_size(MPI_COMM_WORLD, &nproc);
ierr = MPI_Comm_rank(MPI_COMM_WORLD, &proc);

if (proc==0 && (n%nproc != 0 || n%2 != 0)) {
   printf("Number of processes incompatible with the size\n");
   ierr = MPI_Abort(MPI_COMM_WORLD,1);
}
A = (int *) malloc((n/nproc+1) * sizeof(int));

for (i=1; i<= n/nproc; i++)
   A[i]=proc;

printf("proc %i iter=0 A = ", proc);
for (i = 1; i <= n/nproc; i++)
   printf("%i ", A[i]);
printf("\n");

for (iter=1; iter<=n; iter++)
{
  if (proc == 0)
     prev=nproc-1;
  else
     prev=proc-1;
  if (proc==nproc-1)
     next=0;
  else
     next=proc+1;

  if (proc%2 == 0) {
     ierr = MPI_Send(&(A[n/nproc]),1,MPI_INT,next,iter,MPI_COMM_WORLD);
     ierr = MPI_Recv(A,1,MPI_INT,prev,iter,MPI_COMM_WORLD,&status);
  }
  else {
     ierr = MPI_Recv(A,1,MPI_INT,prev,iter,MPI_COMM_WORLD,&status);
     ierr = MPI_Send(&(A[n/nproc]),1,MPI_INT,next,iter,MPI_COMM_WORLD);
  }
  for (i=n/nproc; i>=1; i--)
     A[i]=A[i-1];
  printf("proc %i iter=%i A = ", proc,iter);
  for (i = 1; i <= n/nproc; i++)
     printf("%i ", A[i]);
  printf("\n");
}
free(A);
ierr = MPI_Finalize();

return 0;
}
